lib/web_ui/lib/src/engine/ulps.dart

```diff
@@ -0,0 +1,194 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+part of engine;
+
+// This is a small library to handle stability for floating point operations.
+//
+// Since we are representing an infinite number of real numbers in finite
+// number of bits, when we perform comparisons of coordinates for paths for
+// example, we want to make sure that line and curve sections that are too
+// close to each other (number of floating point numbers
+// representable in bits between two numbers) are handled correctly and
+// don't cause algorithms to fail when we perform operations such as
+// subtraction or between checks.
+//
+// Small introduction into floating point comparison:
+//
+// For some good articles on the topic, see
+// https://randomascii.wordpress.com/category/floating-point/page/2/
+//
+// Here is the 32 bit IEEE representation:
+//   uint32_t mantissa : 23;
+//   uint32_t exponent : 8;
+//   uint32_t sign : 1;
+// As you can see it was carefully designed to be reinterpreted as an integer.
+//
+// Ulps stands for unit in the last place. ulp(x) is the gap between two
+// floating point numbers nearest x.
+
+/// Converts a sign-bit int (float interpreted as int) into a 2s complement
+/// int. Also converts 0x80000000 to 0. Allows result to be compared using
+/// int comparison.
+int signBitTo2sCompliment(int x) => (x & 0x80000000) != 0 ? (-(x & 0x7fffffff)) : x;
+
+/// Convert a 2s compliment int to a sign-bit (i.e. int interpreted as float).
+int twosComplimentToSignBit(int x) {
+  if ((x & 0x80000000) == 0) {
+    return x;
+  }
+  x = ~x + 1;
+  x |= 0x80000000;
+  return x;
+}
+
+class _FloatBitConverter {
+  final Float32List float32List = Float32List(1);
+  late Int32List int32List;
+  _FloatBitConverter() {
+    int32List = float32List.buffer.asInt32List(0, 1);
```

Comment:
`late` has some cost, because it is compiled into a function. If you move the initialization into the initializer list, you can make it `final`, which is free. You may need to use a `factory` though, because the two fields are backed by the same buffer.

Reply:
Done.